[CodeComplete] Compute type relations for global cached results

ahoppen opened the pull request:
Computing the type relation for every item in the code completion cache is way to expensive (~4x slowdown for global completion that imports `SwiftUI`). Instead, compute a type’s supertypes (protocol conformances and superclasses) once and write their USRs to the cache. To compute a type relation we can then check if the contextual type is in the completion item’s supertypes.

This reduces the overhead of computing the type relations (again global completion that imports `SwiftUI`) to ~6% – measured by instructions executed.

Technically, we might miss some conversions like
- retroactive conformances inside another module (because we can’t cache them if that other module isn’t imported)
- complex generic conversions (just too complicated to model using USRs)

Because of this, we never report an `unrelated` type relation for global items but always default to `unknown`.

But I believe this change covers the most common cases and is a good tradeoff between accuracy and performance.

rdar://83846531

---

@rintaro I needed to undo the change you suggested in https://github.com/apple/swift/pull/40999#discussion_r806541085:

When we were only storing AST-based types, we could have a boolean flag to also consider metatypes. This trick is harder to pull off with USR-based types, because we can’t just invoke the MetatypeType constructor with the instance type.

We could hack together a USR-based instance type to Metatype transformation by appending an "m" to the instance type’s USR but I decided to store the metatype as another possible result type of `CodeCompletionResultType` because:
1. Performing USR manipulations when computing type relations seems pretty fragile to me
2. We don’t have an option to store a metatype’s supertypes if we just append an "m" to the USR and thus can’t determine "Convertible" type relations for metatypes (to be fair we can’t do this today either but it should be possible by computing the metatype supertypes in `USRBasedType::fromType`
3. I think we can afford to save another word by making `CodeCompletionResultType` contain `SmallVector<Type, 1>` instead of `Type`.

ahoppen commented:
@swift-ci Please smoke test macOS

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on lib/IDE/CodeCompletionResultType.cpp:
```diff
@@ -11,50 +11,227 @@
+// MARK: - USRBasedTypeContext
+
+USRBasedTypeContext::USRBasedTypeContext(const ExpectedTypeContext *TypeContext,
+                                         const DeclContext *DC,
```
Is `DC` used only for getting `Void` type?
If so, I'd suggest to use `Type::getASTContext()` from one of the expected types. Passing `DC` looks just confusing.

rintaro commented on include/swift/IDE/CodeCompletionResultType.h:
```diff
@@ -107,54 +109,191 @@ enum class CodeCompletionResultTypeRelation : uint8_t {
+/// are represented by the same USR.
+/// A type with an empty USR represents the null type.
+class USRBasedType {
+  std::string USR;
+
+  /// The superclasses of the type and all protocol conformances that are
+  /// declared in the same module that the type was defined.
+  std::vector<const USRBasedType *> Supertypes;
```
This type is allocated on a BumpAllocator, right? If so we can't use non-trivial type fields (std::string and std::vector). Since the allocator doesn't know its contents, it doesn't destruct the objects even if the allocator itself is destructed.

These should be `StringRef` and `ArrayRef<USRBaseType *>` those storage is the same allocator as USRBaseType.

ahoppen replied:
Good catch. I didn’t think about that.

rintaro reviewed (commented):
I'm still thinking how to avoid using vector for result type...

ahoppen commented on lib/IDE/CodeCompletionResultType.cpp:
```diff
@@ -11,50 +11,227 @@
+      for (auto Proto : OpaqueType->getConformsTo()) {
+        USRConforms.push_back(
+            USRBasedType::fromType(Proto->getDeclaredInterfaceType(), Arena));
+      }
```
Which result are you expecting here?
```swift
class MyClass {}
protocol MyProto {}
func fn<T: MyClass & MyProto>() -> T {
  return MyClass() // error: cannot convert return expression of type 'MyClass' to return type 'T'
}
```
When adding `MyClass: MyProto`.
```swift
class MyClass: MyProto {}
protocol MyProto {}
func fn<T: MyClass & MyProto>() -> T { // warning: redundant conformance constraint 'T' : 'MyProto'
  return MyClass() // error: cannot convert return expression of type 'MyClass' to return type 'T'
}
```

rintaro replied:
Hm, OK, we can't know that the actual expected type is if the context type is a generic parameter. But that's same for 
```swift
class MyClass {}
func fn<T: MyProto>() -> T {
  return #^HERE^#
}
```
right? So should we limit this to `OpaqueTypeArchetypeType`?

Anyway, I believe we still need to handle `getSuperclass()` too.
```swift
class C {}
class P {}
class D: C, P {}
class E: P {}
func fn() -> some C & P {
  return #^HERE^#
}
```
D is OK, but E is not.

ahoppen replied:
Yeah, that makes sense. Implemented in https://github.com/apple/swift/pull/41000/commits/10d5d0dd15417a7be1e626f26f9673fa1364e5af

ahoppen commented on lib/IDE/CodeCompletionResultType.cpp:
```diff
@@ -11,50 +11,227 @@
+}
+
+const USRBasedType *USRBasedType::fromType(Type Ty, USRBasedTypeArena &Arena) {
+  // FIXME: We can't mangle archetypes, so we treat them as null USRBasedTypes.
```
This actually works. We are reporting convertible type relations for types that conform to `MyProto` here. I added a test case for it.

rintaro replied:
I mean when suggesting it. The test case would be something like:
```swift
//-- Lib.swift
func returnSomeMyProto() -> some MyProto { ... }

//-- test.swift
import Lib
func foo<T: MyProto>(_: T)
func test() {
  foo(#^HERE^#)
} 
```

ahoppen replied:
I see, I decided to special case `some` with a single protocol for now as representing the conjunction of two types isn’t currently supported in `USRBasedType` and it’s not too simple to put it in. Since the entire USRBasedType thing is just an under-approximation anyway, I think it’s fine if we can’t represent `some MyProto & MyOtherProto` for now. https://github.com/apple/swift/pull/41000/commits/6aff8ef500ee97e10ed55b84a8da5ac298391655

ahoppen commented on include/swift/IDE/CodeCompletionResultType.h:
```diff
@@ -107,54 +109,191 @@ enum class CodeCompletionResultTypeRelation : uint8_t {
+private:
+  /// Cache of the \c Void type because its regularly used in type relation
+  /// computation.
+  const USRBasedType *VoidType;
```
Makes sense. To do this, I needed to pre-compute the `Void` USR type when creating a `USRBasedTypeArena` at which point we don’t have an `ASTContext`. I thus hard-coded the USR of the `Void` type, which should be safe, I think because we don’t expect that to change.

ahoppen commented:
> I'm still thinking how to avoid using vector for result type…

I actually think the vector is a rather nice solution.
- It solves the problem in a pretty elegant way that doesn’t require any USR-manipulations or anything of the sort
- It doesn’t consume too much additional memory (just one `size_t` for the `SmallVector<_, 1>` in the common case that there is only one result type)
- It’s extensible should we find other code completion items that can have multiple result type because there are multiple potential candidates where this might come in handy at some point. I can’t think of any concrete examples but I have a feeling that we might discover more completion results that have such a type duality.

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on include/swift/IDE/CodeCompletionResultType.h:
```diff
@@ -107,54 +109,202 @@ enum class CodeCompletionResultTypeRelation : uint8_t {
+  /// This is not a single unique type because for code completion we consider
+  /// e.g. \c Int as producing both an \c Int metatype and an \c Int instance
+  /// type.
+  SmallVector<PointerUnion<Type, const USRBasedType *>, 1> ResultTypes;
```
I understand that using a "collection" might be a good solution.

However, again, `CodeCompletionResultType` resides in `BumpPtrAllocator` so we can't use `SmallVector` which requires the destructor call. `ArrayRef` might be OK, but how about using `llvm::TrailingObject` so we can pack the `size()` with `IsNotApplicable` flag, and save a pointer size of `data()`. `USRBasedType` could use `TrailingObject` too...

ahoppen replied:
Ah, now I finally understand what your concern about the collection was. As discussed offline, I decided to store the result types using two pointers since we don’t need a collection with more than 2 elements at the moment anyway https://github.com/apple/swift/pull/41000/commits/9ffc3a0543f5f197ccfb116646af7b9976d23bd7

rintaro commented on lib/IDE/CodeCompletionResultType.cpp:
```diff
@@ -11,50 +11,233 @@
+  if (auto Nominal = Ty->getAnyNominal()) {
+    // Sorted conformances so we get a deterministic supertype order and can
+    // assert that USRBasedTypes with the same USR have the same supertypes.
+    auto Conformances = Nominal->getAllConformances(/*sorted=*/true);
```
```
protocol P1 {}
protocol P2: P1 {}

class B: P2 {}

protocol Q1 {}
protocol Q2: Q1 {}

class C: B, Q2 {}
```
In this case, `USRBasedType::forType(C)` would be like
```
p1 = {"P1"}
p2 = {"P2", &p1}
b  = {"B", &p1, &p2}
q1 = {"Q1"}
q2 = {"Q2", &q1}
c  = {"C", &p1, &p2, &q1, &q2, &b}
```
right? This looks a waste of the space. Could you limit `Supertypes` only contains the "direct" super types? So the structure be like:
```
p1 = {"P1"}
p2 = {"P2", &p1}
b  = {"B", &p2}
q1 = {"Q1"}
q2 = {"Q2", &q1}
c  = {"C", &q2, &b}
```
When calculating the type relation, we should be able to collect all super type USRs by recursively look into the graph. Of course we need to be careful not to visit the same node twice, but it shouldn't be hard.

ahoppen replied:
Good idea. It might slow down the computation of USRBasedTypes a little but since that should only happen once to build up the cache, I think it’s a good tradeoff https://github.com/apple/swift/pull/41000/commits/daf8fb056196686f65b42ee0fd4d0fa641c7cc0c

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on lib/IDE/CodeCompletionResultType.cpp:
```diff
@@ -11,50 +11,261 @@
+
+  // For opaque types like 'some View', consider them equivalent to 'View'.
+  if (auto OpaqueType = Ty->getAs<ArchetypeType>()) {
+    if (auto Existential = OpaqueType->getExistentialType()) {
```
Does this support protocol with associated types? Adding a test case would be great.

ahoppen replied:
It does, but while playing around with them I discovered another bug
```swift
public protocol ProtoWithAssocType {
  associatedtype MyAssoc
}


public func makeProtoWithAssocType() -> some ProtoWithAssocType { return StructWithAssocType() }

func protoWithAssocTypeInGenericContext<T: ProtoWithAssocType>() -> T {
   return #^COMPLETE^#
 }
```
would suggest `makeProtoWithAssocType` as “Identical” even though it is not valid here. The problem was that we were doing some of the special opaque return type handling for USRBasedTypes for all archetypes. Fix was to only apply them to `OpaqueTypeArchetypeType`. This regressed test cases which are using a generic context, but I think that’s fine since the main known limitation of the USR-based type comparisons is that they don’t support generic contexts. https://github.com/apple/swift/pull/41000/commits/5a1dd49d99acc3da122f20eb8fef81c3b44b94a6

rintaro commented on lib/IDE/CodeCompletionResultType.cpp:
```diff
@@ -11,50 +11,261 @@
+    return TypeRelation::Identical;
+  }
+  for (auto *Supertype : ResultType->getSupertypes()) {
+    if (this->typeRelation(Supertype, VoidType) >= TypeRelation::Convertible) {
```
When the hierarchy contains diamond conformance,
```swift
protocol Base {}
protocol P1: Base {}
protocol P2: Base {}
struct S: P1, P2 {}
```
I think `Base` is visited twice. I'm not sure this happens lot, and not sure it causes excessive waste. But this logic should have "visited" check.

ahoppen replied:
Added it https://github.com/apple/swift/pull/41000/commits/b12d82a3e811a22aedadb06a9a46af20517be664

rintaro commented on include/swift/IDE/CodeCompletionResultType.h:
```diff
@@ -107,54 +109,248 @@ enum class CodeCompletionResultTypeRelation : uint8_t {
+                           PointerUnion<Type, const USRBasedType *> ResultType1,
+                           PointerUnion<Type, const USRBasedType *> ResultType2)
+      : ResultType1AndIsApplicable(ResultType1, IsApplicable),
+        ResultType2(ResultType2) {}
```
Insert `assert(ResultType1.is<Type>() == ResultType2.is<Type>());` or do we have scenario mixing them?

ahoppen replied:
We don’t have a scenario where we’re mixing them but we never make assumptions that both of them are of the same kind either. Adding the assertion just adds an artificial restriction that’s not necessary IMO.

rintaro reviewed (approved):
Added some comment, but nothing critical. We can do them in followups

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SouceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen committed:
[CodeComplete] Compute type relations for global cached results

Computing the type relation for every item in the code completion cache is way to expensive (~4x slowdown for global completion that imports `SwiftUI`). Instead, compute a type’s supertypes (protocol conformances and superclasses) once and write their USRs to the cache. To compute a type relation we can then check if the contextual type is in the completion item’s supertypes.

This reduces the overhead of computing the type relations (again global completion that imports `SwiftUI`) to ~6% – measured by instructions executed.

Technically, we might miss some conversions like
- retroactive conformances inside another module (because we can’t cache them if that other module isn’t imported)
- complex generic conversions (just too complicated to model using USRs)

Because of this, we never report an `unrelated` type relation for global items but always default to `unknown`.

But I believe this change covers the most common cases and is a good tradeoff between accuracy and performance.

rdar://83846531

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen committed:
[CodeCompletion] Ignore Sendable conformances

Sendable conformances are lazily synthesized as they are needed by the compiler. Depending on whether we checked whether a type conforms to Sendable before constructing the USRBasedType, we get different results for its conformance, causing assertion failures. For now, always ignore the Sendable conformance. We need to come up with a better solution for this problem in the future.

ahoppen committed:
[CodeCompletion] Check USRBasedTypeCache earlier when converting AST-based types

Previously, for every type, we were reconstructing its entire supertype hierarchy as USRBasedTypes before checking the already computed USRBasedTypes. This lead to unneccary duplicated work.

ahoppen commented:
@swift-ci Please smoke test

hamishknight commented:
https://ci.swift.org/job/oss-swift-pr-test-macoss/123 started failing after this change, reverting in https://github.com/apple/swift/pull/41627
